### PR TITLE
Less redundant Reshape spec

### PR DESCRIPTION
**Pull request by taliesinb**
The core ability this gives is to allow certain kinds of reshape without requiring you to know all the dimensions up-front, in a more flexible way than the current codes 0 and -1. This make it easier for humans to express certain kinds of ops, but also makes it possible for machines to express certain reshapes without having to know exact sizes of irrelevant parts of the input. 

That's what I'm using this for our fork: helping our neural network framework unroll recurrent networks more easily and efficiently. You typically can prove ahead of time that an unrolling algorithm is correct, and for speed reasons you do NOT want to track tensor dimensions during unrolling at run-time. But Reshape layer demands to know all tensor sizes even when the desired resize operation doesn't actually depend on them.

So this PR adds 3 more codes to Reshape that let it figure out more things itself:
* -2: copy all remaining dimensions from input shape to output shape
* -3: merge two adjacent dimensions in input shape
* -4: split a single dimension in input shape into two (where either one can be inferred if given the other).

It should be pretty straightforward to add more codes in future, like maybe a code for merging all remaining dimensions. 

These are documented formally in the description of "target_shape" with proper indices etc, the previous documentation was very vague.

Here's an example that illustrates all three: let's say you want to 'broadcast' an operation across a sequence by simply merging the sequence and batch dimensions together and then doing an ordinary batched op. That kind of merge op can be expressed as a Reshape with target `(-3, -2)` -- merge the first two dimensions, then copy the rest. After you apply the op you'll want to do an unmerge: which is `(-4,slen,-1,-2)`. Notice that the only static information we needed to know here was the sequence length to unmerge properly. The dimensions and rank of the inner tensors was not needed. 

Another example: any existing code will be able to replace any sequence of "0s" with a single "-2".

### CODE REVIEW

**Comment by msollami** [comment]
oops i just meant to like

**Comment by piiswrong** [comment]
This is useful but looks really complicated. maybe add some examples in the doc?

@winstywang @gengyifeng @tornadomeet @sxjscience please have a look and evaluate if this is easy enough to understand

**Comment by taliesinb** [comment]
@piiwrong I can definitely do that. Where should the examples go? In the description field of shape? Or elsewhere?

**Comment by taliesinb** [comment]
BTW, the right way to think about this thing is as a little automata or machine that walks the input dimensions and the target shape simultaneously and produces the output dimensions. Which is already how it worked before -- that's what `0` did, for example -- it copied the corresponding dim from the input dims to the output dims.

Now it gets some more operations, like taking two input dims and only emitting one output dim, or taking one input dim and emitting two output dims. That way the rank of your output can be different from the rank of your input, but the way it chooses the dimensions is guaranteed to be valid as long as the rest of the spec is also valid.

So if you just imagine this little machine walking along the input dims and writing to the output dims it's quite easy to 'read' the instructions in your head.

**Comment by piiswrong** [comment]
I think examples should go reshape.cc in describe

**Comment by taliesinb** [comment]
@piiswrong Ok I added examples for each code. Let me know if you want me to reformat or tweak that description in any way. I could add a line of English for some of those examples if particular ones are unclear.

**Comment by sxjscience** [comment]
In the unmerging part, why do we need to use (-4, seqlen, -1, -2) instead of (-4, seqlen, -2)?

**Comment by taliesinb** [comment]
@sxjscience Your syntax `(-4, seqlen, -2)` would require a special case, the two 'arguments' to Split2 currently have to be ordinary positive integers (although ONE of them can be -1, to mean that it is equal to the original input length divided by the other integer). 

I can add that case for `-2`, but to me it only really makes sense in the second position to Split2, like what you wrote. The other way round, `(-4, -2, seqlen)`, would 'mix up' the higher dimensions in a weird non-local way. Do any of other codes make sense as arguments to Split2? Let's think. `0` would just copy the input dimension you're splitting, so the other dimension would be forced to be `1`. That's pretty useless. `-3` and `-4` seem pretty complicated to support and I'm trying to think what kind of interpretation these have. 

So maybe I can summarize these thoughts as: I've defined a machine that operates on stacks of dimensions and pops and pushes them. This machine is useful because it lets us write simple 'programs' that decide the output shape and they work against many input shapes in a uniform way. This is good because we don't need to have say 3 or 4 new kinds of layers that hard-code these behaviors when a flexible spec can give us the same thing. 

But your question is really asking why this machine doesn't allow 'recursion' (embedding a code inside another code). And I think the answer is: I don't know what recursion should really do in this machine, and I don't think it would make the machine any more powerful. 

What do you think?

**Comment by sxjscience** [comment]
@taliesinb Sorry I've misunderstood the meaning of -4. So there must be two numbers after -4 and one of them can be -1. I have one more question. Could we use multiple -4s in the shape argument to help inference the target shape? Like `(-4, a, -1, -4, b, -1)`?

**Comment by taliesinb** [comment]
@sxjscience Yes. I've just added an example (the last one) that shows using two -4 codes, where each one has a -1 arg, like your example.

**Comment by piiswrong** [comment]
This to be sure: does this change the default behavior in any case?

**Comment by taliesinb** [comment]
@piiswrong: any 'shape' that currently is valid will continue to be valid and have the same output. 

Some shapes that used to be invalid are now valid. For example, the shape `(3,0)` used to be invalid -- there was a check to prevent you manually setting a dimension (3) AND using the 0 code at the same time. 

There was nothing wrong such such a thing in principle, but I guess it was prevented because there wasn't a model for exactly what it should mean. The "stack machine" model that I've described is such a model, and furthermore it actually *wants* you to be able to combine codes like `0` with fixed dimensions like `3`, so therefore I've removed that check.

TLDR: the design is 100% backward compatible.

**Comment by sxjscience** [comment]
LGTM

**Comment by taliesinb** [comment]
@sxjscience Yeah this will cut out a lot of code in our framework that were just there to redundantly track tensor dims for the purposes of feeding them to Reshape layer. I'm very excited about this.
